hudi-spark-datasource/hudi-spark/src/test/scala/org/apache/hudi/functional/TestSevenToEightUpgrade.scala

```diff
@@ -102,4 +110,260 @@ class TestSevenToEightUpgrade extends RecordLevelIndexTestBase {
+    // ------------------------------------------------------------------
+    println("== Step 3: Updating records ==")
+    val updateDF = Seq(
+      (1, "Alice", 2000, 15),  // update Alice
+      (2, "Bob",   2000, 25)   // update Bob
+    ).toDF("id", "name", "ts", "price")
+
+    updateDF.write.format("hudi")
+      .options(hudiOptions)
+      .mode("append")
+      .save(basePath)
+
+    // ------------------------------------------------------------------
+    // Step 4: Trigger compaction so that a new base file is produced.
+    // For inline compaction, we set inline compaction options.
+    // ------------------------------------------------------------------
+    val compactionOptions = hudiOptions ++ Map(
+      HoodieCompactionConfig.INLINE_COMPACT.key -> "true",
+      HoodieCompactionConfig.INLINE_COMPACT_NUM_DELTA_COMMITS.key -> "1"
+    )
+    // Writing the same updateDF triggers inline compaction in a MERGE_ON_READ table.
+    updateDF.write.format("hudi")
+      .options(compactionOptions)
+      .mode("append")
+      .save(basePath)
+
+    // ------------------------------------------------------------------
+    // Step 5: Perform a clean operation.
+    // ------------------------------------------------------------------
+    val cleanOptions = hudiOptions ++ Map(
+      HoodieCleanConfig.CLEANER_FILE_VERSIONS_RETAINED.key -> "1"
+    )
+    val client = DataSourceUtils.createHoodieClient(
+      spark.sparkContext, "", basePath, tableName, cleanOptions.asJava
+    ).asInstanceOf[SparkRDDWriteClient[HoodieRecordPayload[Nothing]]]
+    // Trigger cleaning via an additional write (which invokes cleaning)
+    val cleanInstant = client.scheduleTableService(Option.empty(), TableServiceType.CLEAN)
+    client.clean(cleanInstant.get())
+    client.close()
+
+    // ------------------------------------------------------------------
+    // Step 6: Upgrade to version 8.
+    // ------------------------------------------------------------------
+    val upgradedOptions = hudiOptions ++ Map(
+      HoodieWriteConfig.AUTO_UPGRADE_VERSION.key -> "true",
+      HoodieWriteConfig.WRITE_TABLE_VERSION.key -> HoodieTableVersion.EIGHT.versionCode().toString
+    )
+    metaClient = HoodieTableMetaClient.reload(metaClient)
+    new UpgradeDowngrade(metaClient, getWriteConfig(upgradedOptions, basePath), context, SparkUpgradeDowngradeHelper.getInstance)
+      .run(HoodieTableVersion.EIGHT, null)
+    // assert table version
+    metaClient = HoodieTableMetaClient.reload(metaClient)
+    assertEquals(HoodieTableVersion.EIGHT, metaClient.getTableConfig.getTableVersion)
+
+    // TODO: remove this once we have the fix from PR#13009
+    val props = metaClient.getTableConfig.getProps
+    props.setProperty(HoodieTableConfig.RECORD_MERGE_MODE.key(), RecordMergeMode.EVENT_TIME_ORDERING.name())
+    HoodieTableConfig.update(metaClient.getStorage, metaClient.getMetaPath, props)
```

Comment:
Had to set the merge mode in addition to payload class. Just setting payload class does not automatically set the merge mode in upgrade step. I hope that will be fixed by #13009

Reply:
ack. 
CC @yihua : when you work on upgrade handler merge mode issue, can you take care of this